syncstorage/storage/sql/queries_spanner.py

```diff
@@ -0,0 +1,121 @@
+"""
+
+COLLECTION_CURRENT_TIMESTAMP = """\
+SELECT CURRENT_TIMESTAMP() as now, last_modified FROM user_collections
+WHERE userid=@userid AND collection=@collectionid
+"""
+
+STORAGE_SIZE = """\
+SELECT SUM(payload_size) FROM (
+    SELECT CHAR_LENGTH(payload) as payload_size, userid FROM bso
+    WHERE userid=@userid AND ttl > CURRENT_TIMESTAMP()
+) GROUP BY userid
+"""
+
+COLLECTIONS_SIZES = """\
+SELECT collection, SUM(payload_size) FROM (
+    SELECT CHAR_LENGTH(payload) as payload_size, collection FROM bso
+    WHERE userid=@userid AND ttl > CURRENT_TIMESTAMP()
+) GROUP BY collection
+"""
+
+COLLECTIONS_COUNTS = """
+SELECT collection, COUNT(collection) FROM bso WHERE userid=@userid AND
+ttl > CURRENT_TIMESTAMP()
+GROUP BY collection
+"""
+
+
+def FIND_ITEMS(bso, params):
+    """Item search query.
+
+    Unlike all the other pre-built queries, this one really can't be written
+    as a simple string.  We need to include/exclude various WHERE clauses
+    based on the values provided at runtime.
+    """
+    fields = params.get("fields", None)
+    if fields is None:
+        query = select(col for col in bso.columns
+                       if col.name != "payload_size")
+    else:
+        query = select([bso.c[field] for field in fields])
+    query = query.where(bso.c.userid == bindparam("userid"))
+    query = query.where(bso.c.collection == bindparam("collectionid"))
+    # Filter by the various query parameters.
+    if "ids" in params:
+        # Sadly, we can't use a bindparam in an "IN" expression.
+        query = query.where(bso.c.id.in_(params.get("ids")))
+    if "newer" in params:
+        query = query.where(bso.c.modified > bindparam("newer"))
+    if "newer_eq" in params:
+        query = query.where(bso.c.modified >= bindparam("newer_eq"))
+    if "older" in params:
+        query = query.where(bso.c.modified < bindparam("older"))
+    if "older_eq" in params:
+        query = query.where(bso.c.modified <= bindparam("older_eq"))
+    if "ttl" in params:
+        query = query.where(bso.c.ttl > bindparam("ttl"))
+    # Sort it in the order requested.
+    # We always sort by *something*, so that limit/offset work consistently.
+    # The default order is by timestamp, which is efficient due to the index.
+    # We also want to sort by "id" as a secondary column to ensure a consistent
+    # total ordering, which is important when paginating large requests.
+    # Since "id" is in the primary key, that *should* be efficient...
+    sort = params.get("sort", None)
+    if sort == 'index':
+        order_args = [bso.c.sortindex.desc(), bso.c.id.desc()]
+    elif sort == 'oldest':
+        order_args = [bso.c.modified.asc(), bso.c.id.asc()]
+    else:
+        order_args = [bso.c.modified.desc(), bso.c.id.asc()]
+    # ...but unfortunately, sorting by "id" causes a significant slowdown on
+    # older versions of MySQL, so it's disabled by default and must be
+    # explicitly opted in to via config.
```

Comment:
You can probably drop this mysql-specific workaround for spanner.

Reply:
Indeed.